cast/src/lib.rs

```diff
@@ -1209,6 +1225,75 @@ impl SimpleCast {
 
         Ok(code)
     }
+
+    /// Etherscan can either return one raw string that includes all of the solidity for a verified
+    /// contract or a json struct surrounded in an extra set of {} that includes a directory
+    /// structure with paths and source code.
+    fn get_sources_from_etherscan_source_value(
+        contract_name: &str,
+        etherscan_source: &str,
+    ) -> Result<Vec<(String, String)>> {
+        if etherscan_source.len() >= 4 &&
+            &etherscan_source[0..2] == "{{" &&
+            &etherscan_source[etherscan_source.len() - 2..] == "}}"
+        {
+            let json = &etherscan_source[1..etherscan_source.len() - 1];
+            let parsed: EtherscanSourceJsonMetadata = serde_json::from_str(&json)?;
+            Ok(parsed
+                .sources
+                .into_iter()
+                .map(|(path, source_struct)| (path, source_struct.content))
+                .collect())
+        } else {
+            Ok(vec![(contract_name.to_string(), etherscan_source.to_string())])
+        }
+    }
+
+    /// Fetches the source code of verified contracts from etherscan and expands the resulting
+    /// files to a directory for easy perusal.
+    /// ```
+    /// # use cast::SimpleCast as Cast;
+    /// # use ethers_core::types::Chain;
+    /// # use std::path::PathBuf;
+    ///
+    /// # async fn expand() -> eyre::Result<()> {
+    ///      Cast::expand_etherscan_source_to_directory(Chain::Mainnet, "0xBB9bc244D798123fDe783fCc1C72d3Bb8C189413".to_string(), "<etherscan_api_key>".to_string(), PathBuf::from("output_dir")).await?;
+    /// #    Ok(())
+    /// # }
+    /// ```
+    pub async fn expand_etherscan_source_to_directory(
+        chain: Chain,
+        contract_address: String,
+        etherscan_api_key: String,
+        output_directory: PathBuf,
+    ) -> eyre::Result<()> {
+        let client = Client::new(chain, etherscan_api_key)?;
+        let meta = client.contract_source_code(contract_address.parse()?).await?;
+        if meta.items.iter().any(|item| item.abi == "Contract source code not verified") {
```

Comment:
we should handle this while deserializing the response in the etherscan client and make this an Option. Mind opening an issue for this on ethers or even a PR?